src/compute/src/render/join/mz_join_core.rs

```diff
@@ -576,31 +575,27 @@ where
     }
 
     /// Process keys until at least `fuel` output tuples produced, or the work is exhausted.
-    fn work<L, I, YFn, C>(
+    fn work<L, CB: ContainerBuilder, YFn>(
         &mut self,
-        output: &mut OutputHandleCore<T, CapacityContainerBuilder<C>, Tee<T, C>>,
+        output: &mut OutputHandleCore<T, EffortBuilder<CB::Container>, Tee<T, CB::Container>>,
         mut logic: L,
         yield_fn: YFn,
         work: &mut usize,
     ) where
-        I: IntoIterator<Item = D>,
-        L: FnMut(C1::Key<'_>, C1::Val<'_>, C2::Val<'_>) -> I,
+        L: FnMut(C1::Key<'_>, C1::Val<'_>, C2::Val<'_>, &mut JoinSession<T, CB, CB::Container>),
         YFn: Fn(usize) -> bool,
-        C: PushContainer,
-        (D, T, Diff): PushInto<C>,
+        <CB as ContainerBuilder>::Container: ContainerBuilder,
```

Comment:
This constraint looks off, a container should not be a builder. Probably caused by the `EffortBuilder<CB::Container>` type further up.